contracts/utils/convex/ConvexStakingWrapperYield.sol

```diff
@@ -0,0 +1,111 @@
+pragma solidity 0.8.6;
+pragma experimental ABIEncoderV2;
+
+import './ConvexStakingWrapper.sol';
+import "@yield-protocol/utils-v2/contracts/access/AccessControl.sol";
+
+struct Balances {
+    uint128 art; // Debt amount
+    uint128 ink; // Collateral amount
+}
+
+interface ICauldron {
+    /// @dev Each vault records debt and collateral balances_.
+    function balances(bytes12 vault) external view returns (Balances memory);
+}
+
+//Staking wrapper for Yield platform
+//use convex LP positions as collateral while still receiving rewards
+contract ConvexStakingWrapperYield is ConvexStakingWrapper,AccessControl {
+    using SafeERC20 for IERC20;
+    using Address for address;
+
+
+    mapping(address=>bytes12[]) public vaults;
+    address cauldron;
+
+    constructor() public {}
+
+    function initialize(
+        address _curveToken,
+        address _convexToken,
+        address _convexPool,
+        uint256 _poolId,
+        address _vault
+    ) external override auth{
+        require(!isInit, 'already init');
+        owner = address(0xa3C5A1e09150B75ff251c1a7815A07182c3de2FB); //TODO: Find why this needs to be set to convex multisig
+        emit OwnershipTransferred(address(0), owner);//TODO: Find why this needs to be done
+        _tokenname = string(abi.encodePacked('Staked ', ERC20(_convexToken).name(), ' Yield'));
+        _tokensymbol = string(abi.encodePacked('stk', ERC20(_convexToken).symbol(), '-yield'));
+        isShutdown = false;
+        isInit = true;
+        curveToken = _curveToken;
+        convexToken = _convexToken;
+        convexPool = _convexPool;
+        convexPoolId = _poolId;
+        collateralVault = _vault; //TODO: Add the join address
+
+        //add rewards
+        addRewards();
+        setApprovals();
+    }
+
+    function setCauldron(address _cauldron) external auth{
+        require(_cauldron!=address(0), 'cauldron address cannot be 0');
+        cauldron = _cauldron;
+    }
+
+    // Set the locations of vaults where the user's funds have been deposited & the accounting is kept
+    function setVault(address _account, bytes12 _vault) external auth {
+        bytes12[] storage userVault = vaults[_account];
+        for (uint256 i = 0; i < userVault.length; i++) {
+            require(userVault[i] != _vault, 'already added');
+        }
+        userVault.push(_vault);
+        vaults[_account] = userVault;
+    }
+
+    function removeVault(address _account, bytes12 _vault) external auth {
+        bytes12[] storage userVault = vaults[_account];
+        for (uint256 i = 0; i < userVault.length; i++) {
+            if(userVault[i] == _vault){
+                vaults[_account] = remove(i,userVault);
+                break;
+            }
+        }
+    }
+
+    function remove(uint _index,bytes12[] storage userVault) internal returns (bytes12[] memory){
+        require(_index < userVault.length, "index out of bound");
+
+        for (uint i = _index; i < userVault.length - 1; i++) {
```

Comment:
Instead of this very expensive loop (1 SSTORE for position after _index), if `_index` is not the last position in the array, copy the value of the last position to `_index` and then `pop`. You lose the ordering, but we don't care about that.